src/main/java/org/victorrobotics/frc/dtlib/drivetrain/swerve/trajectory/DTSwerveTrajectoryGenerator.java

```diff
@@ -0,0 +1,280 @@
+package org.victorrobotics.frc.dtlib.drivetrain.swerve.trajectory;
+
+import org.victorrobotics.frc.dtlib.drivetrain.DTAccelerationLimit;
+import org.victorrobotics.frc.dtlib.drivetrain.DTVelocityLimit;
+import org.victorrobotics.frc.dtlib.drivetrain.swerve.trajectory.DTSwerveTrajectory.Point;
+
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Rotation2d;
+
+public class DTSwerveTrajectoryGenerator {
+    private static final double SMOOTH_TOLERANCE = 0.005;
+
+    private static final DTAccelerationLimit DEFAULT_ACCEL_LIMIT = new DTAccelerationLimit();
+    private static final DTVelocityLimit     DEFAULT_VELO_LIMIT  = new DTVelocityLimit();
+
+    private static final double DEFAULT_DISTANCE_BETWEEN_POINTS = 0.05;
+    private static final double DEFAULT_SMOOTH_WEIGHT           = 0.75;
+    private static final double DEFAULT_VELOCITY_POWER          = 0.2;
+    private static final double DEFAULT_VELOCITY_COEFFICIENT    = 5;
+
+    private DTAccelerationLimit accelerationLimit;
+    private DTVelocityLimit     velocityLimit;
+
+    private double distanceBetweenPoints;
+    private double smoothWeight;
+    private double velocityPower;
+    private double velocityCoefficient;
+
+    public DTSwerveTrajectoryGenerator() {
+        distanceBetweenPoints = DEFAULT_DISTANCE_BETWEEN_POINTS;
+        smoothWeight = DEFAULT_SMOOTH_WEIGHT;
+        velocityPower = DEFAULT_VELOCITY_POWER;
+        velocityCoefficient = DEFAULT_VELOCITY_COEFFICIENT;
+
+        accelerationLimit = DEFAULT_ACCEL_LIMIT;
+        velocityLimit = DEFAULT_VELO_LIMIT;
+    }
+
+    public void setAccelerationLimit(DTAccelerationLimit limit) {
+        accelerationLimit = limit;
+    }
+
+    public void setVelocityLimit(DTVelocityLimit limit) {
+        velocityLimit = limit;
+    }
+
+    public void setDistanceBetweenPoints(double distanceMeters) {
+        distanceBetweenPoints = distanceMeters;
+    }
+
+    public void setSmoothWeight(double weight) {
+        smoothWeight = weight;
+    }
+
+    public void setVelocityPower(double power) {
+        velocityPower = power;
+    }
+
+    public void setVelocityCoefficient(double coefficient) {
+        velocityCoefficient = coefficient;
+    }
+
+    public DTSwerveTrajectory generate(Pose2d... waypoints) {
+        double[][] interpolatedPoints = interpolate(waypoints);
+        double[][] smoothPoints = smooth(interpolatedPoints);
+
+        Pose2d[] path = new Pose2d[smoothPoints.length];
+        for (int i = 0; i < path.length; i++) {
+            path[i] = new Pose2d(smoothPoints[i][0], smoothPoints[i][1],
+                    Rotation2d.fromDegrees(smoothPoints[i][2]));
+        }
+
+        // cumulative time, cumulative distance, commanded velocity, predicted
+        // velocity, linear acceleration, curvature
+        double[][] dataValues = new double[6][path.length];
```

Comment:
Consider making a simple wrapper class for the data values rather than an array of 6 doubles. A simple "data class" with 6 public double members would be ok. You could use it within the Point class as well. Alternatively, use enumeration constants for the indexes.

Reply:
I left it as a 2D array to maximize performance before, but creating a temporary mutable wrapper would definitely make it more clear.